### ISSUE TITLE
Allow Exceptions to be passed on from the server to the client

### ISSUE THREAD

**Issue by vlidholt** [serverpod/serverpod#486]
From @AndryHTC

Is it possible to catch **Domain Errors** thrown by the server, from client? Would be super useful to retrieve the error that caused the fail of the client call.

i.e.
## Server
```
Future createUser(String username) {
    ...
    throw UserAlreadyTaken()
    ...
}
```

## Client
```
try {
    await client.endpoint.createUser('username');
} on UserAlreadyTaken {
    // show SnackBar and suggest alternative usernames
}
``` 

Could potentially work by overriding a custom Exception class and/or adding exceptions to the yaml files.

**Comment by AndryHTC** [comment]
We can't wait for it! 😃 Will be a game changer: would feel like working entirely with Flutter forgetting about there's a back-end behind. A true Flutter backend... So again the claim **The missing server for Flutter** will be affirmed again and again 🚀

**Comment by damienissa** [comment]
@vlidholt Hi. I already have some ideas about the implementation, can I try to implement them and make a Pull Request?

**Comment by yahu1031** [comment]
@damienissa Sure. That would be great.

**Comment by vlidholt** [comment]
@damienissa please post an outline of what you are planning before getting started, so we are on the same side before getting started. :)

**Comment by damienissa** [comment]
@vlidholt  @yahu1031  Thanks for the quick response. 
I already have a working proof of concept but it needs to be improved.

We can throw any classes from or endpoints. That means that we can use ```SerializableEntity``` as an exception.

The next step is adding ```on SerizlizableEntity catch (entity)``` in the ```handleUriCall```.

After that, we need to add a new type that is a child of the ```Result``` 
like ```ServerExceptionResult``` with the ```entity``` field with type ```SerializableEntity```

The last step for the server side is adding a custom response structure if the result is a ```ServerExceptionResult```:
```dart
{
    'className': result.entity.runtimeType.toString(),
    'data': result.entity.toJson(),
    'exception': true
}
```

Also, we need to change a little bit file generator to update ```Protocol``` generation to add the ability to parse and throw a custom exception on a client side like this:
```dart
if (data is Map<String, dynamic> &&
        data["className"] != null &&
        data["exception"] == true) {
  throw deserializeByClassName(data);
}
```
But it's only a concept)

**Comment by damienissa** [comment]
@vlidholt  Also, similar behavior needs to be added for void returning functions.

**Comment by damienissa** [comment]
@vlidholt After researches I understand that ```throw``` objects must be a subtype of ```Exception```. 
If ignore this aspect regular ```} catch (e) {``` will not catch it. Needs always wright ```} on MyCustomException catch (e) {```

I have added an ability to create a class that will subtype of Exception. ```Yaml``` structure very similar to ```class```:
```yaml
exception: UserNotFound
fields:
  userId: String
  errorMessage: String
```

Now it works fine =)

**Comment by vlidholt** [comment]
@damienissa, this looks like a good approach. My thinking was that the `UserNotFound` exceptions should extend a new `ServerpodException` (that in turn extends `Exception`. If a `ServerpodException` is caught when processing the call, it returns a 500 status code, but the body is empty. The body could instead contain the serialized `UserNotFound` exception, or if the body is empty a standardized `ServerpodException`. Makes sense?

**Comment by damienissa** [comment]
@vlidholt It sounds good. I'll update my code. And I'll back to you.

**Comment by vlidholt** [comment]
Thank you! 🤩

**Comment by damienissa** [comment]
@vlidholt PR is ready for review. I'll be waiting for your proposals.

**Comment by vlidholt** [comment]
Thank you! Very nice, I'm swamped with work right now, but will review in the next few days!

**Comment by klkucaj** [comment]
@damienissa thank you for this feature, i think this is one of the most important feature a BE server should have.
@vlidholt btw, i am waiting for this and for all closed ones to be released, you guys made a lot of progress, THANK YOU!
But this feature we really need it because there are a lot of exception a BE server can throw:
UserNotFound
OperationNotAllowed
InvalidCredentials
AccessDenied
Unauthorized
.....

**Comment by damienissa** [comment]
@vlidholt  @yahu1031 
When you planned to release this cool feature?

**Comment by yahu1031** [comment]
Once merged, you can have this in hit cli version

**Comment by vlidholt** [comment]
It will be released with Serverpod 1.1, we are adding a few more features, but we are almost there!

**Comment by vlidholt** [comment]
I'm closing this issue as it is implemented on `main`
